Kernel/ACPI/MultiProcessorParser.cpp

```diff
@@ -0,0 +1,268 @@
+        p_floating_pointer_str += 16;
+    }
+    return nullptr;
+}
+MultiProcessor::FloatingPointer* MultiProcessorParser::search_floating_pointer_in_bios_area()
+{
+    auto floating_pointer_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)0xE0000)), PAGE_ROUND_UP(0xFFFFF - 0xE0000), "MultiProcessor Parser floating_pointer Finding #2", Region::Access::Read, false, true);
+    char* p_floating_pointer_str = (char*)(PhysicalAddress(0xE0000).as_ptr());
+    for (char* floating_pointer_str = (char*)floating_pointer_region->vaddr().offset(offset_in_page((u32)(0xE0000))).as_ptr(); floating_pointer_str < (char*)(floating_pointer_region->vaddr().offset(offset_in_page((u32)(0xE0000))).get() + (0xFFFFF - 0xE0000)); floating_pointer_str += 16) {
+#ifdef MUTLIPROCESSOR_DEBUG
+        dbg() << "MultiProcessor: Looking for floating pointer structure in BIOS area @ V0x " << String::format("%x", floating_pointer_str) << ", P0x" << String::format("%x", p_floating_pointer_str);
+#endif
+        if (!strncmp("_MP_", floating_pointer_str, strlen("_MP_")))
+            return (MultiProcessor::FloatingPointer*)p_floating_pointer_str;
+        p_floating_pointer_str += 16;
+    }
+    return nullptr;
+}
+
+Vector<unsigned>& MultiProcessorParser::get_pci_bus_ids()
+{
+    Vector<unsigned>& pci_bus_ids = *new Vector<unsigned>();
+    for (auto* entry : m_bus_entries) {
+        auto entry_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)entry)), PAGE_ROUND_UP(m_configuration_table_length), "MultiProcessor Parser Parsing Bus Entry", Region::Access::Read, false, true);
+        auto* v_entry_ptr = (MultiProcessor::BusEntry*)entry_region->vaddr().offset(offset_in_page((u32)entry)).as_ptr();
+        if (!strncmp("PCI   ", v_entry_ptr->bus_type, strlen("PCI   ")))
+            pci_bus_ids.append(v_entry_ptr->bus_id);
+    }
+    return pci_bus_ids;
+}
+
+MultiProcessorParser& MultiProcessorParser::the()
+{
+    ASSERT(MultiProcessorParser::is_initialized());
+    return *s_parser;
+}
+
+Vector<RefPtr<PCIInterruptOverrideMetadata>>& MultiProcessorParser::get_pci_interrupt_redirections()
+{
+    dbg() << "MultiProcessor: Get PCI IOAPIC redirections";
+    Vector<RefPtr<PCIInterruptOverrideMetadata>>& overrides = *new Vector<RefPtr<PCIInterruptOverrideMetadata>>();
+    Vector<unsigned>& pci_bus_ids = get_pci_bus_ids();
+    for (auto* entry : m_io_interrupt_redirection_entries) {
+
+        auto entry_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)entry)), PAGE_ROUND_UP(m_configuration_table_length), "MultiProcessor Parser Parsing Bus Entry", Region::Access::Read, false, true);
+        auto* v_entry_ptr = (MultiProcessor::IOInterruptAssignmentEntry*)entry_region->vaddr().offset(offset_in_page((u32)entry)).as_ptr();
+        dbg() << "MultProcessor: Parsing Entry P 0x" << String::format("%x", entry) << ", V 0x" << String::format("%x", v_entry_ptr);
```

Comment:
I think you can write this as:
```suggestion
        dbg() << "MultiProcessor: Parsing Entry P " << entry << ", V 0x" << v_entry_ptr;
```

Reply:
I'll fix the typo and remove the String::format() since we don't need it, but shouldn't we output addresses with hexadecimal notation (e.g. using "0x" before the number) ?

Reply:
My mistake, I see that the PrintfImplementation with `dbg()` already prints out hexadecimal prefix :)

Reply:
@awesomekling The `entry` variable isn't pointer, therefore I still need to call `String::format()` on it, but `v_entry_ptr` is indeed a pointer so no need to call `String::format()` for it.